planner/core/plan.go

```diff
@@ -282,6 +282,12 @@ type LogicalPlan interface {
 	// predicateSimplification consolidates different predcicates on a column and its equivalence classes.
 	predicateSimplification(opt *logicalOptimizeOp) LogicalPlan
 
+	// constantPropagation generate new constant predicate according to column equivalence relation
+	constantPropagation(parentPlan LogicalPlan, currentChildIdx int, opt *logicalOptimizeOp) (newRoot LogicalPlan)
+
+	// recursiveGetConstantPredicates recursive find constant predicate, used for the constant propagation rule
+	recursiveFindAndPullUpConstantPredicates() []expression.Expression
```

Comment:
Since these 2 methods are only used for this rule, so how about implementing them as 2 pure statistics functions like `constantPropagation(currentPlan, parentPlan LogicalPlan, currentChildIdx int, opt *logicalOptimizeOp) (newRoot LogicalPlan)` on `rule_constant_propagation.go`? 
There are already too many methods attached on `LogicalPlan`, which makes it hard to maintain...

Reply:
There is a lot of logic in the rule that needs to be recursively implemented layer by layer. 
Different logical plans will have different logic. I feel that abstracting it into an interface is more cohesive at the architectural level.
BTW, a base class can inherit multiple interfaces. For example, the interface of statistics, and the interface of different rules. But I don't know if golang can support it.